File: ios/DutchLearnerPackage/Sources/DutchLearnerPackage/Utilities/SRSAlgorithm.swift
//
//  SRSAlgorithm.swift
//  DutchLearner
//
//  SM-2 Algorithm Implementation for Spaced Repetition
//  Based on SuperMemo 2 algorithm by Piotr Wozniak
//

import Foundation

public enum SRSAlgorithm {
    // Constants
    public static let initialEaseFactor: Double = 2.5
    public static let minEaseFactor: Double = 1.3
    public static let graduatingInterval: Int = 1 // First successful review: 1 day
    public static let easyInterval: Int = 4 // "Easy" first review: 4 days
    public static let secondInterval: Int = 6 // Second successful review: 6 days

    /// Initialize SRS data for a new card
    public static func initializeCard() -> SRSCardData {
        return SRSCardData(
            interval: 0,
            repetitions: 0,
            easeFactor: initialEaseFactor,
            dueDate: Date(), // Due immediately
            lastReviewDate: nil,
            reviewHistory: [],
            isNew: true
        )
    }

    /// Convert answer correctness and hints used to SM-2 quality rating (0-5)
    /// - Parameters:
    ///   - isCorrect: Whether the answer was correct
    ///   - hintsUsed: Number of hints used (0 = no hints)
    /// - Returns: SRSQuality rating
    public static func getQuality(isCorrect: Bool, hintsUsed: Int = 0) -> SRSQuality {
        if !isCorrect {
            // Failed - quality 0-2 based on how many hints were used
            if hintsUsed == 0 { return .totalBlackout } // Complete failure
            if hintsUsed == 1 { return .incorrectButRemembered } // Failed with one hint
            return .incorrectButEasyRecall // Failed with multiple hints
        }

        // Correct answer - quality 3-5 based on hints
        if hintsUsed >= 2 { return .correctWithDifficulty } // Correct but with difficulty (multiple hints)
        if hintsUsed == 1 { return .correctWithHesitation } // Correct with some hesitation (one hint)
        return .perfectRecall // Perfect recall (no hints)
    }

    /// Calculate the next review interval and ease factor based on SM-2 algorithm
    /// - Parameters:
    ///   - cardData: Current SRS card data
    ///   - quality: Quality rating (0-5)
    /// - Returns: Updated SRS card data
    public static func calculateNextReview(cardData: SRSCardData, quality: SRSQuality) -> SRSCardData {
        let now = Date()
        var interval = cardData.interval
        var repetitions = cardData.repetitions
        var easeFactor = cardData.easeFactor

        // Update ease factor based on quality (SM-2 formula)
        // EF' = EF + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02))
        let q = Double(quality.rawValue)
        let newEaseFactor = max(
            minEaseFactor,
            easeFactor + (0.1 - (5.0 - q) * (0.08 + (5.0 - q) * 0.02))
        )

        let newInterval: Int
        let newRepetitions: Int

        // If quality < 3, the item needs to be reviewed again (failed)
        if !quality.isPassing {
            // Reset to learning phase
            newInterval = 0
            newRepetitions = 0
        } else {
            // Successful review
            if repetitions == 0 {
                // First successful review
                newInterval = graduatingInterval
                newRepetitions = 1
            } else if repetitions == 1 {
                // Second successful review
                newInterval = secondInterval
                newRepetitions = 2
            } else {
                // Subsequent reviews: interval = previous interval * ease factor
                newInterval = Int(round(Double(interval) * newEaseFactor))
                newRepetitions = repetitions + 1
            }
        }

        // Calculate due date
        let dueDate = Calendar.current.date(byAdding: .day, value: newInterval, to: now) ?? now

        // Create review record
        let reviewRecord = ReviewRecord(
            date: now,
            quality: quality,
            interval: newInterval
        )

        var newReviewHistory = cardData.reviewHistory
        newReviewHistory.append(reviewRecord)

        return SRSCardData(
            interval: newInterval,
            repetitions: newRepetitions,
            easeFactor: newEaseFactor,
            dueDate: dueDate,
            lastReviewDate: now,
            reviewHistory: newReviewHistory,
            isNew: false
        )
    }

    /// Update SRS card data after a review
    /// - Parameters:
    ///   - cardData: Current SRS card data (or nil for new card)
    ///   - isCorrect: Whether the answer was correct
    ///   - hintsUsed: Number of hints used
    /// - Returns: Updated SRS card data
    public static func updateCard(cardData: SRSCardData?, isCorrect: Bool, hintsUsed: Int = 0) -> SRSCardData {
        // Initialize new card if needed
        let currentData = cardData ?? initializeCard()

        // Convert performance to quality rating
        let quality = getQuality(isCorrect: isCorrect, hintsUsed: hintsUsed)

        // Calculate next review
        return calculateNextReview(cardData: currentData, quality: quality)
    }

    /// Check if a card is due for review
    public static func isCardDue(_ cardData: SRSCardData, now: Date = Date()) -> Bool {
        return cardData.dueDate <= now
    }

    /// Get the number of days until the card is due
    public static func daysUntilDue(_ cardData: SRSCardData, now: Date = Date()) -> Int {
        let components = Calendar.current.dateComponents([.day], from: now, to: cardData.dueDate)
        return components.day ?? 0
    }

    /// Categorize a card based on its interval
    public static func getCardCategory(_ cardData: SRSCardData) -> CardCategory {
        if cardData.isNew { return .new }
        if cardData.interval == 0 { return .learning }
        if cardData.interval <= 21 { return .young }
        return .mature
    }

    /// Calculate retention rate from review history
    public static func calculateRetentionRate(_ reviewHistory: [ReviewRecord]) -> Double {
        guard !reviewHistory.isEmpty else { return 100.0 }

        let successfulReviews = reviewHistory.filter { $0.quality.isPassing }.count
        return (Double(successfulReviews) / Double(reviewHistory.count)) * 100.0
    }
}

public enum CardCategory {
    case new
    case learning
    case young
    case mature

    public var displayName: String {
        switch self {
        case .new: return "New"
        case .learning: return "Learning"
        case .young: return "Young"
        case .mature: return "Mature"
        }
    }
}

File: src/utils/srsAlgorithm.ts
import type { SRSCardData, SRSQuality, ReviewRecord } from "../types/srs";

/**
 * SM-2 Algorithm Implementation for Spaced Repetition
 * Based on SuperMemo 2 algorithm by Piotr Wozniak
 */

// Constants
const INITIAL_EASE_FACTOR = 2.5;
const MIN_EASE_FACTOR = 1.3;
const GRADUATING_INTERVAL = 1; // First successful review: 1 day
const EASY_INTERVAL = 4; // "Easy" first review: 4 days
const MILLISECONDS_PER_DAY = 24 * 60 * 60 * 1000;

/**
 * Initialize SRS data for a new card
 */
export function initializeSRSCard(): SRSCardData {
  return {
    interval: 0,
    repetitions: 0,
    easeFactor: INITIAL_EASE_FACTOR,
    dueDate: Date.now(), // Due immediately
    lastReviewDate: null,
    reviewHistory: [],
    isNew: true,
  };
}

/**
 * Convert answer correctness and hints used to SM-2 quality rating (0-5)
 *
 * @param isCorrect - Whether the answer was correct
 * @param hintsUsed - Number of hints used (0 = no hints)
 * @returns SRSQuality rating (0-5)
 */
export function getQualityFromPerformance(
  isCorrect: boolean,
  hintsUsed: number = 0,
): SRSQuality {
  if (!isCorrect) {
    // Failed - quality 0-2 based on how many hints were used
    if (hintsUsed === 0) return 0; // Complete failure
    if (hintsUsed === 1) return 1; // Failed with one hint
    return 2; // Failed with multiple hints
  }

  // Correct answer - quality 3-5 based on hints
  if (hintsUsed >= 2) return 3; // Correct but with difficulty (multiple hints)
  if (hintsUsed === 1) return 4; // Correct with some hesitation (one hint)
  return 5; // Perfect recall (no hints)
}

/**
 * Calculate the next review interval and ease factor based on SM-2 algorithm
 *
 * @param cardData - Current SRS card data
 * @param quality - Quality rating (0-5)
 * @returns Updated SRS card data
 */
export function calculateNextReview(
  cardData: SRSCardData,
  quality: SRSQuality,
): SRSCardData {
  const now = Date.now();
  let { interval, repetitions, easeFactor } = cardData;

  // Update ease factor based on quality (SM-2 formula)
  // EF' = EF + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02))
  const newEaseFactor = Math.max(
    MIN_EASE_FACTOR,
    easeFactor + (0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)),
  );

  let newInterval: number;
  let newRepetitions: number;

  // If quality < 3, the item needs to be reviewed again (failed)
  if (quality < 3) {
    // Reset to learning phase
    newInterval = 0;
    newRepetitions = 0;
  } else {
    // Successful review
    if (repetitions === 0) {
      // First successful review
      newInterval = GRADUATING_INTERVAL;
      newRepetitions = 1;
    } else if (repetitions === 1) {
      // Second successful review
      newInterval = 6;
      newRepetitions = 2;
    } else {
      // Subsequent reviews: interval = previous interval * ease factor
      newInterval = Math.round(interval * newEaseFactor);
      newRepetitions = repetitions + 1;
    }
  }

  // Calculate due date
  const dueDate = now + newInterval * MILLISECONDS_PER_DAY;

  // Create review record
  const reviewRecord: ReviewRecord = {
    date: now,
    quality,
    interval: newInterval,
  };

  return {
    interval: newInterval,
    repetitions: newRepetitions,
    easeFactor: newEaseFactor,
    dueDate,
    lastReviewDate: now,
    reviewHistory: [...cardData.reviewHistory, reviewRecord],
    isNew: false,
  };
}

/**
 * Update SRS card data after a review
 *
 * @param cardData - Current SRS card data (or undefined for new card)
 * @param isCorrect - Whether the answer was correct
 * @param hintsUsed - Number of hints used
 * @returns Updated SRS card data
 */
export function updateSRSCard(
  cardData: SRSCardData | undefined,
  isCorrect: boolean,
  hintsUsed: number = 0,
): SRSCardData {
  // Initialize new card if needed
  const currentData = cardData || initializeSRSCard();

  // Convert performance to quality rating
  const quality = getQualityFromPerformance(isCorrect, hintsUsed);

  // Calculate next review
  return calculateNextReview(currentData, quality);
}

/**
 * Check if a card is due for review
 *
 * @param cardData - SRS card data
 * @param now - Current timestamp (defaults to Date.now())
 * @returns True if the card is due for review
 */
export function isCardDue(
  cardData: SRSCardData,
  now: number = Date.now(),
): boolean {
  return cardData.dueDate <= now;
}

/**
 * Get the number of days until the card is due
 *
 * @param cardData - SRS card data
 * @param now - Current timestamp (defaults to Date.now())
 * @returns Days until due (negative if overdue)
 */
export function getDaysUntilDue(
  cardData: SRSCardData,
  now: number = Date.now(),
): number {
  const millisUntilDue = cardData.dueDate - now;
  return Math.ceil(millisUntilDue / MILLISECONDS_PER_DAY);
}

/**
 * Categorize a card based on its interval
 *
 * @param cardData - SRS card data
 * @returns 'new' | 'learning' | 'young' | 'mature'
 */
export function getCardCategory(
  cardData: SRSCardData,
): "new" | "learning" | "young" | "mature" {
  if (cardData.isNew) return "new";
  if (cardData.interval === 0) return "learning";
  if (cardData.interval <= 21) return "young";
  return "mature";
}

/**
 * Calculate retention rate from review history
 *
 * @param reviewHistory - Array of review records
 * @returns Retention rate as a percentage (0-100)
 */
export function calculateRetentionRate(reviewHistory: ReviewRecord[]): number {
  if (reviewHistory.length === 0) return 100;

  const successfulReviews = reviewHistory.filter(
    (record) => record.quality >= 3,
  ).length;
  return Math.round((successfulReviews / reviewHistory.length) * 100);
}
